# Before we start

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
import pandas as pd
pd.set_option('display.width', 180)


In [8]:
from google.colab import output
from IPython.display import Javascript

def resize_colab_cell():
    display(Javascript('google.colab.output.setIframeHeight(0, true, {maxHeight: 5000})'))

get_ipython().events.register('pre_run_cell', resize_colab_cell)


In [2]:
import pandas as pd

# Replace 'path/to/homelessness.csv' with the path to your CSV file
df = pd.read_csv('/content/drive/MyDrive/wk10_pandas_learner/Resources/data.csv', index_col = 0)

# Print the first 5 rows of the DataFrame
print(df.head())


          Pulse  Maxpulse  Calories
Duration                           
60          110       130     409.1
60          117       145     479.0
60          103       135     340.0
45          109       175     282.4
45          117       148     406.0


index_col is a parameter of the pandas.read_csv function that specifies which column of the CSV file to use as the row labels of the DataFrame1. If index_col is set to 0, it means that the first column of the CSV file will be used as the row labels. For example, if the CSV file looks like this:

As we do not need to do it for our data so we shall remove the index_col parameter but we shall need to use it later.

* * *

#Viewing the Data
----------------

One of the most used method for getting a quick overview of the DataFrame, is the `head()` method.

The `head()` method returns the headers and a specified number of rows, starting from the top.





In [3]:
import pandas as pd

# Replace 'path/to/homelessness.csv' with the path to your CSV file
df = pd.read_csv('/content/drive/MyDrive/wk10_pandas_learner/Resources/data.csv')

# Print the first 5 rows of the DataFrame
print(df.head(10))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


In our examples we will be using a CSV file called 'data.csv'.

Download [data.csv](https://www.w3schools.com/python/pandas/data.csv), or open [data.csv](https://www.w3schools.com/python/pandas/data.csv.txt) in your browser.

**Note:** if the number of rows is not specified, the `head()` method will return the top 5 rows.

### Example

Print the first 5 rows of the DataFrame:
```
import pandas as pd

# Replace 'path/to/homelessness.csv' with the path to your CSV file
df = pd.read_csv('/content/drive/MyDrive/WK10-pandas/Resources/data.csv')

# Print the first 5 rows of the DataFrame
print(df.head())
```
[Try it Yourself »](https://www.w3schools.com/python/pandas/trypandas.asp?filename=demo_pandas_head)

* * *





In [4]:
import pandas as pd

# Replace 'path/to/homelessness.csv' with the path to your CSV file
df = pd.read_csv('/content/drive/MyDrive/wk10_pandas_learner/Resources/data.csv')

# Print the first 5 rows of the DataFrame
print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


There is also a `tail()` method for viewing the _last_ rows of the DataFrame.

The `tail()` method returns the headers and a specified number of rows, starting from the bottom.

* * *

In [5]:
print(df.tail())

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


Info About the Data
-------------------

The DataFrames object has a method called `info()`, that gives you more information about the data set.

### Example
```
#Print information about the data:

print(df.info())
```


In [6]:
#Print information about the data:

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


Result
------
```
  <class 'pandas.core.frame.DataFrame'>
  RangeIndex: 169 entries, 0 to 168
  Data columns (total 4 columns):
   #   Column    Non-Null Count  Dtype  
  ---  ------    --------------  -----  
   0   Duration  169 non-null    int64  
   1   Pulse     169 non-null    int64  
   2   Maxpulse  169 non-null    int64  
   3   Calories  164 non-null    float64
  dtypes: float64(1), int64(3)
  memory usage: 5.4 KB
  None
  ```  

[Try it Yourself »](https://www.w3schools.com/python/pandas/trypandas.asp?filename=demo_pandas_info)

Result Explained
----------------

The result tells us there are 169 rows and 4 columns:
```
  RangeIndex: 169 entries, 0 to 168
  Data columns (total 4 columns):
```
And the name of each column, with the data type:
```
   #   Column    Non-Null Count  Dtype  
  ---  ------    --------------  -----  
   0   Duration  169 non-null    int64  
   1   Pulse     169 non-null    int64  
   2   Maxpulse  169 non-null    int64  
   3   Calories  164 non-null    float64
```
Null Values
-----------

The `info()` method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called _cleaning data_, and you will learn more about that in the next chapters.

* * *

Below is an exercise for you that incorporates the specified pandas methods (`to_string`, `head`, `max_rows`, `read_csv`, and `pd.options.display.max_rows`):

**Exercise: Exploring Fitness Tracker Data with pandas**

**Step 1: Import pandas and load the data**

1. Go to **Runtime** > **Change runtime type** and select **Python 3**.
2. Click the **Files** icon in the left sidebar and upload the CSV file with the fitness tracker data.
3. Open a new code cell and import the `pandas` library:

   ```python
   import pandas as pd
   ```

4. Read the CSV file into a pandas DataFrame using `read_csv()`:

   ```python
   filename = "your_file_name.csv"  # Replace with the actual file name
   df = pd.read_csv(filename)
   ```

5. (Optional) To limit the number of rows displayed by default, set `pd.options.display.max_rows`:

   ```python
   pd.options.display.max_rows = 10  # This will show only the first 10 rows
   ```

**Step 2: Examine the DataFrame**

1. Use `head()` to view the first few rows:

   ```python
   print(df.head())
   ```

2. Check the DataFrame's size and data types using `info()`:

   ```python
   df.info()
   ```

**Step 3: Get descriptive statistics**

1. Use `describe()` to summarize numerical columns (Duration, Pulse, Maxpulse, Calories):

   ```python
   print(df.describe())
   ```

**Step 4: Access specific data**

1. Access column values by name:

   ```python
   durations = df["Duration"]
   calories = df["Calories"]
   ```

2. Access rows by index (zero-based):

   ```python
   first_row = df.iloc[0]  # Gets the first row
   last_row = df.iloc[-1]  # Gets the last row
   ```

3. Access rows by label (if your Dataframe has an index set):

   ```python
   # Assuming you have set an index, e.g., df.set_index("Date")
   row_with_index_2023_11_20 = df.loc["2023-11-20"]
   ```

4. Access rows using boolean indexing:

   ```python
   high_calories = df[df["Calories"] > 400]  # Select rows with calories exceeding 400
   ```

**Step 5: Convert the DataFrame to string**

1. Use `to_string()` to create a string representation of the DataFrame:

   ```python
   df_string = df.to_string()
   print(df_string)
   ```

**Challenge:**

1. Find the row with the longest exercise duration.
2. Calculate the average calorie burn per exercise duration.
3. Group the data by exercise duration and find the average pulse for each group.
4. Create a visualization (e.g., histogram, scatter plot) using Matplotlib or Seaborn to explore the relationships between variables.

Remember to replace `"your_file_name.csv"` with the actual name of your uploaded CSV file. As you explore and manipulate the data, feel free to experiment with other pandas methods and functions to gain deeper insights and practice your data analysis skills!

In [16]:
import pandas as pd

filename = '/content/drive/MyDrive/wk10_pandas_learner/Resources/data.csv'
df = pd.read_csv(filename, index_col=0)

print(df.head())

df.info()

print(df.describe())

<IPython.core.display.Javascript object>

          Pulse  Maxpulse  Calories
Duration                           
60          110       130     409.1
60          117       145     479.0
60          103       135     340.0
45          109       175     282.4
45          117       148     406.0
<class 'pandas.core.frame.DataFrame'>
Index: 169 entries, 60 to 75
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pulse     169 non-null    int64  
 1   Maxpulse  169 non-null    int64  
 2   Calories  164 non-null    float64
dtypes: float64(1), int64(2)
memory usage: 5.3 KB
            Pulse    Maxpulse     Calories
count  169.000000  169.000000   164.000000
mean   107.461538  134.047337   375.790244
std     14.510259   16.450434   266.379919
min     80.000000  100.000000    50.300000
25%    100.000000  124.000000   250.925000
50%    105.000000  131.000000   318.600000
75%    111.000000  141.000000   387.600000
max    159.000000  184.000000  1860.400000
